src/charm.py

```diff
@@ -905,7 +905,7 @@ def _hosts(self) -> set:
         ]
         return set(hosts)
 
-    @property
+    @cached_property
```

Comment:
sounds to good to be true - Is there any case where you need to invalidate the cache while in the hook, e.g. change in IP, or planned units within a hook?